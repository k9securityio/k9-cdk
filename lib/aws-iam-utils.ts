import {Effect, PolicyDocument, PolicyStatement} from "@aws-cdk/aws-iam";

/**
 * Gets the unique set of Principal ARNs (or tokenized representation) that appear in the Principal element of
 * a Statement that Allows access from an existing PolicyDocument.  Parallels K9PolicyFactory#getAllowedPrincipalArns.
 *
 * Limitations:
 *  * does not examine the statement's condition element
 *  * does not do anything with NotPrincipal
 *
 * @param policyDocument to analyze
 * @return the set of allowed principal ARNs or tokens
 */
export function getAllowedPrincipalArns(policyDocument: PolicyDocument): Set<string> {
    const origStatements = new Array<PolicyStatement>();
    const origAllowedAWSPrincipals = new Set<string>();
    if (policyDocument.statementCount > 0) {
        const origPolicyJSON: any = policyDocument.toJSON();
        for (let statementJson of origPolicyJSON.Statement) {
            let origStatement = PolicyStatement.fromJson(statementJson);
            origStatements.push(origStatement);
            if (origStatement.effect == Effect.ALLOW &&
                origStatement.hasPrincipal) {
                let origStatementJSON = origStatement.toStatementJson();
                console.log(`origStatementJSON: ${JSON.stringify(origStatementJSON)}`);
                if (origStatementJSON?.Principal?.AWS) {
                    let awsPrincipals = origStatementJSON.Principal.AWS;
                    if (typeof awsPrincipals == 'string') {
                        console.log(`origStatementJSON.Principal (str): ${awsPrincipals}`);
                        origAllowedAWSPrincipals.add(awsPrincipals)
                    } else if (Array.isArray(awsPrincipals)) {
                        console.log(`origStatementJSON.Principal (array): ${awsPrincipals}`);
                        awsPrincipals.forEach(function (value) {
                            origAllowedAWSPrincipals.add(value);
                        });
                    } else {
                        console.log(`origStatementJSON.Principal (${typeof awsPrincipals}): ${JSON.stringify(awsPrincipals)}`);
                    }
                }
            }
        }
    }
    return origAllowedAWSPrincipals;
}