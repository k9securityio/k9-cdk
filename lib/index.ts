export import k9policy = require('./k9policy');
export import s3 = require('./s3');
