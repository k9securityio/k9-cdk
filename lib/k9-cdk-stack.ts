import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3'
import {BucketPolicy} from '@aws-cdk/aws-s3'
import {AnyPrincipal, Effect, PolicyStatement, PolicyStatementProps} from "@aws-cdk/aws-iam";
import {readFileSync} from 'fs';

export type ArnEqualsTest = "ArnEquals"

export type ArnLikeTest = "ArnLike";

export type ArnConditionTest =
    | ArnEqualsTest
    | ArnLikeTest;

export enum AccessCapability {
    AdministerResource = "administer-resource",
    ReadData = "read-data",
    WriteData = "write-data",
    DeleteData = "delete-data",
}

export interface K9AccessSpec {
    accessCapability: AccessCapability
    allowPrincipalArns: Set<string>
    test: ArnConditionTest
}

export interface K9DesiredAccessSpecs {
    // K9DesiredAccessSpecs may be able to replace K9AccessCapabilities in the near future
    [accessCapability: string]: K9AccessSpec;
}

export class K9AccessCapabilities {

    constructor(
        readonly allowAdministerResourceArns?: Set<string>,
        readonly allowAdministerResourceTest?: ArnConditionTest,
        readonly allowReadDataArns?: Set<string>,
        readonly allowReadDataTest?: ArnConditionTest,
        readonly allowWriteDataArns?: Set<string>,
        readonly allowWriteDataTest?: ArnConditionTest,
        readonly allowDeleteDataArns?: Set<string>,
        readonly allowDeleteDataTest?: ArnConditionTest,
    ) {

    }

    // ??? Add support for custom actions
    // will probably encourage users to create custom statements directly using the policy instead of trying to model
}

export interface K9BucketPolicyProps extends s3.BucketPolicyProps {
    readonly k9AccessCapabilities: K9AccessCapabilities
    readonly bucket: s3.Bucket
}

export class K9PolicyFactory {

    SUPPORTED_CAPABILITIES = new Array<AccessCapability>(
        AccessCapability.AdministerResource,
        AccessCapability.ReadData,
        AccessCapability.WriteData,
        AccessCapability.DeleteData,
    );

    SUPPORTED_SERVICES = new Set<string>(["S3"]);

    _K9CapabilityMapJSON: Object = JSON.parse(readFileSync('./lib/capability_summary.json').toString());
    K9CapabilityMapByService: Map<string, Object> = new Map(Object.entries(this._K9CapabilityMapJSON));

    getAccessSpec(accessCapability: AccessCapability, desiredCapabilities: K9AccessCapabilities): K9AccessSpec {
        switch (accessCapability) {
            case "administer-resource":
                return {
                    accessCapability: accessCapability,
                    allowPrincipalArns: desiredCapabilities.allowAdministerResourceArns ? desiredCapabilities.allowAdministerResourceArns : new Set<string>(),
                    test: desiredCapabilities.allowAdministerResourceTest ? desiredCapabilities.allowAdministerResourceTest : "ArnEquals"
                };
            case "read-data":
                return {
                    accessCapability: accessCapability,
                    allowPrincipalArns: desiredCapabilities.allowReadDataArns ? desiredCapabilities.allowReadDataArns : new Set<string>(),
                    test: desiredCapabilities.allowReadDataTest ? desiredCapabilities.allowReadDataTest : "ArnEquals"
                };
            case "write-data":
                return {
                    accessCapability: accessCapability,
                    allowPrincipalArns: desiredCapabilities.allowWriteDataArns ? desiredCapabilities.allowWriteDataArns : new Set<string>(),
                    test: desiredCapabilities.allowWriteDataTest ? desiredCapabilities.allowWriteDataTest : "ArnEquals"
                };
            case "delete-data":
                return {
                    accessCapability: accessCapability,
                    allowPrincipalArns: desiredCapabilities.allowDeleteDataArns ? desiredCapabilities.allowDeleteDataArns : new Set<string>(),
                    test: desiredCapabilities.allowDeleteDataTest ? desiredCapabilities.allowDeleteDataTest : "ArnEquals"
                };
            default:
                throw Error(`unsupported capability: ${accessCapability}`)
        }
    }


    makeBucketPolicy(scope: cdk.Construct, id: string, props: K9BucketPolicyProps): BucketPolicy {

        const policy = new s3.BucketPolicy(scope, `${id}Policy`, {bucket: props.bucket});

        for (let accessCapability of this.SUPPORTED_CAPABILITIES) {
            let accessSpec = this.getAccessSpec(accessCapability, props.k9AccessCapabilities);
            let statement = makeAllowStatement(`Restricted-${accessCapability}`,
                this.getActions('S3', accessCapability),
                accessSpec.allowPrincipalArns,
                accessSpec.test);
            policy.document.addStatements(statement)

        }

        policy.document.addStatements(new PolicyStatement({
                effect: Effect.DENY,
                principals: [new AnyPrincipal()],
                actions: ['*'],
                resources: [`${props.bucket.bucketArn}/*`],
                conditions: {
                    Bool: {'aws:SecureTransport': false},
                },
            })
        );

        policy.document.validateForResourcePolicy();
        console.log("validated resource policy");

        return policy
    }

    private getActions(service: string, accessCapabiilty: AccessCapability): Array<string> {
        if (!this.SUPPORTED_SERVICES.has(service) && this.K9CapabilityMapByService.has(service)) {
            throw Error(`unsupported service: ${service}`)
        }

        let serviceCapabilitiesObj: Object = this.K9CapabilityMapByService.get(service) || {};
        let serviceCapabilitiesMap = new Map<string, Array<string>>(Object.entries(serviceCapabilitiesObj));

        let accessCapabilityName = accessCapabiilty.toString();
        if (serviceCapabilitiesMap &&
            serviceCapabilitiesMap.has(accessCapabilityName)) {
            return serviceCapabilitiesMap.get(accessCapabilityName) || Array<string>();
        } else {
            return new Array<string>();
        }
    }
}

function makeAllowStatement(sid: string, actions: Array<string>, arns: Set<string>, test: ArnConditionTest) {
    let policyStatementProps:PolicyStatementProps = {
        sid: sid,
        effect: Effect.ALLOW
    };
    let statement = new PolicyStatement(policyStatementProps);
    statement.addActions(...actions);
    statement.addAnyPrincipal();
    statement.addAllResources();
    statement.addCondition(test, {'aws:PrincipalArn': [...arns]});
    return statement;
}

export class K9CdkStack extends cdk.Stack {
    // TODO - Remove stack definition.  We're not going to vend a stack.
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

    }
}