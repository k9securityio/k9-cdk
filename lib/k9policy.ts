import {AnyPrincipal, ArnPrincipal, Effect, PolicyStatement, PolicyStatementProps} from "@aws-cdk/aws-iam";
import {readFileSync} from 'fs';
import {stringify} from "querystring";

export type ArnEqualsTest = "ArnEquals"

export type ArnLikeTest = "ArnLike";

export type ArnConditionTest =
    | ArnEqualsTest
    | ArnLikeTest;

export enum AccessCapability {
    AdministerResource = "administer-resource",
    ReadConfig = "read-config",
    ReadData = "read-data",
    WriteData = "write-data",
    DeleteData = "delete-data",
}

export interface AccessSpec {
    accessCapabilities: Set<AccessCapability> | AccessCapability
    allowPrincipalArns: Set<string>
    test?: ArnConditionTest
}

export class K9PolicyFactory {

    SUPPORTED_SERVICES = new Set<string>([
        "S3",
        "KMS",
    ]);
    _K9CapabilityMapJSON: Object = JSON.parse(readFileSync(`${__dirname}/capability_summary.json`).toString());
    K9CapabilityMapByService: Map<string, Object> = new Map(Object.entries(this._K9CapabilityMapJSON));
    
    getActions(service: string, accessCapability: AccessCapability): Array<string> {
        if (!this.SUPPORTED_SERVICES.has(service) && this.K9CapabilityMapByService.has(service)) {
            throw Error(`unsupported service: ${service}`)
        }

        let serviceCapabilitiesObj: Object = this.K9CapabilityMapByService.get(service) || {};
        let serviceCapabilitiesMap = new Map<string, Array<string>>(Object.entries(serviceCapabilitiesObj));

        let accessCapabilityName = accessCapability.toString();
        if (serviceCapabilitiesMap &&
            serviceCapabilitiesMap.has(accessCapabilityName)) {
            return serviceCapabilitiesMap.get(accessCapabilityName) || Array<string>();
        } else {
            return new Array<string>();
        }
    }

    _mergeAccessSpecs(target: AccessSpec, addition: AccessSpec) {
        for (let desiredPrincipalArn of addition.allowPrincipalArns) {
            target.allowPrincipalArns.add(desiredPrincipalArn);
        }
        if (target.test) {
            //ok, user has specified a test at some point; ensure this desiredAccessSpec.test matches
            if (target.test != addition.test) {
                let msg = 'Cannot merge AccessSpecs; test attributes do not match:' +
                    `\n${stringify(target)}\n${stringify(addition)}`;
                throw Error(msg);
            }
        } else {
            //first explicit test preference wins
            if(addition.test){
                target.test = addition.test;
            }
        }

    }

    makeAllowStatements(serviceName: string,
                        supportedCapabilities: Array<AccessCapability>,
                        desiredAccess: Array<AccessSpec>,
                        resourceArns: Array<string>): Array<PolicyStatement> {
        let policyStatements = new Array<PolicyStatement>();
        let accessSpecsByCapability: Map<AccessCapability, AccessSpec> = new Map<AccessCapability, AccessSpec>();

        // 1. populate accessSpecsByCapability with fresh AccessSpecs for each supported capability
        // 2. iterate through desiredAccess specs and merge data into what we'll use
        //    important: detect mismatched test types
        //     we can leave `test` unset in the default access specs
        //     and copy the value from the spec being merged if it is set
        //     throw Error on mismatch
        // 3. generate an Allow statement for each supported capability

        for (let supportedCapability of supportedCapabilities) {
            //generate a default access spec for each of the service's supported capabilities
            let effectiveAccessSpec: AccessSpec = {
                accessCapabilities: supportedCapability,
                allowPrincipalArns: new Set<string>(),
                // leave 'test' property unset; will populate from user-provided data
            };
            accessSpecsByCapability.set(supportedCapability, effectiveAccessSpec);

            //now... merge in the user's desired access for this capability
            for (let desiredAccessSpec of desiredAccess) {
                if(desiredAccessSpec.accessCapabilities instanceof Set){
                    for (let desiredCapability of desiredAccessSpec.accessCapabilities) {
                        if (supportedCapability == desiredCapability) {
                            this._mergeAccessSpecs(effectiveAccessSpec, desiredAccessSpec);
                        }
                    }
                } else if (typeof desiredAccessSpec.accessCapabilities == 'string') {
                    if (supportedCapability == desiredAccessSpec.accessCapabilities) {
                        this._mergeAccessSpecs(effectiveAccessSpec, desiredAccessSpec);
                    }
                } else {
                    throw Error(`Unhandled type of accessCapabilities for ${desiredAccessSpec.accessCapabilities}`)
                }
            }
        }

        // ok, time to actually make Allow Statements from our AccessSpecs
        for (let supportedCapability of supportedCapabilities) {

            let accessSpec: AccessSpec = accessSpecsByCapability.get(supportedCapability) ||
                {   //satisfy compiler; should never happen, because we populate at the beginning.
                    //generate a default access spec if none was provided
                    accessCapabilities: new Set([supportedCapability]),
                    allowPrincipalArns: new Set<string>(),
                    test: "ArnEquals"
                }
            ;
            
            let arnConditionTest = accessSpec.test || "ArnEquals";

            let statement = this.makeAllowStatement(`Allow Restricted ${supportedCapability}`,
                this.getActions(serviceName, supportedCapability),
                accessSpec.allowPrincipalArns,
                arnConditionTest,
                resourceArns);
            policyStatements.push(statement);
        }
        return policyStatements;
    }

    makeAllowStatement(sid: string,
                       actions: Array<string>,
                       principalArns: Set<string>,
                       test: ArnConditionTest,
                       resources: Array<string>): PolicyStatement {
        let policyStatementProps: PolicyStatementProps = {
            sid: sid,
            effect: Effect.ALLOW
        };
        let statement = new PolicyStatement(policyStatementProps);
        statement.addActions(...actions);
        statement.addAnyPrincipal();
        statement.addResources(...resources);
        statement.addCondition(test, {'aws:PrincipalArn': [...principalArns].sort()});
        return statement;
    }

    wasLikeUsed(accessSpecs: AccessSpec[]): boolean {
        for (let accessSpec of accessSpecs) {
            if ("ArnLike" == accessSpec.test) {
                return true;
            }
        }
        return false;
    }

    getAllowedPrincipalArns(accessSpecs: AccessSpec[]): Set<string> {
        let allowedPrincipalArns = new Set<string>();
        for (let accessSpec of accessSpecs) {
            accessSpec.allowPrincipalArns.forEach(function (value) {
                allowedPrincipalArns.add(value);
            });
        }
        return allowedPrincipalArns;
    }

    /**
     * k9 wants to deny all AWS accounts and IAM principals not explicitly allowed; this *should*
     * be straightforward, but it isn't because of the way aws-cdk merges and manipulates Principals.
     * @return list of principals for a DenyEveryoneElse statement
     */
    makeDenyEveryoneElsePrincipals(): ArnPrincipal[] {
        /**
         * We should be able to provide AnyPrincipal once (of course), but AWS CDK converts:
         * "Principal": {
         *   "AWS": "*"    // identifies all AWS accounts and IAM.
         * }
         * to:
         * "Principal": "*"  // identifies all principals including AWS Service principals
         *
         * That's a greater scope than we want.
         *
         * So provide AnyPrincipal twice, so aws-cdk maintains the array form.
         * 
         * AWS rewrites the AWS member of the policy on save so
         * only the unique set of principals are included
         * So after these machinations, we end up with what we want.
         */
        return [new AnyPrincipal(), new AnyPrincipal()];
    }

}

