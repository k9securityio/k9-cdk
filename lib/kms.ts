import * as iam from "@aws-cdk/aws-iam";
import {AccountRootPrincipal, Effect, PolicyDocument, PolicyStatement} from "@aws-cdk/aws-iam";
import {AccessCapability, AccessSpec, K9PolicyFactory} from "./k9policy";
import * as cdk from "@aws-cdk/core";

export interface K9KeyPolicyProps {
    readonly k9DesiredAccess: Array<AccessSpec>
}

let SUPPORTED_CAPABILITIES = new Array<AccessCapability>(
    AccessCapability.AdministerResource,
    AccessCapability.ReadConfig,
    AccessCapability.ReadData,
    AccessCapability.WriteData,
    AccessCapability.DeleteData,
);


export function makeKeyPolicy(scope: cdk.Construct, id: string, props: K9KeyPolicyProps): PolicyDocument {
    const policyFactory = new K9PolicyFactory();
    const policy = new iam.PolicyDocument();

    const resourceArns = ['*'];

    const allowStatements = policyFactory.makeAllowStatements("KMS",
        SUPPORTED_CAPABILITIES,
        props.k9DesiredAccess,
        resourceArns);
    policy.addStatements(...allowStatements);

    const denyEveryoneElseStatement = new PolicyStatement({
        sid: 'DenyEveryoneElse',
        effect: Effect.DENY,
        principals: [new AccountRootPrincipal()],
        actions: ['kms:*'],
        resources: resourceArns
    });
    const denyEveryoneElseTest = policyFactory.wasLikeUsed(props.k9DesiredAccess) ?
        'ArnNotLike' :
        'ArnNotEquals';
    const allAllowedPrincipalArns = policyFactory.getAllowedPrincipalArns(props.k9DesiredAccess);
    denyEveryoneElseStatement.addCondition(denyEveryoneElseTest, {
        'aws:PrincipalArn': [...allAllowedPrincipalArns]
    });

    policy.addStatements(
        new PolicyStatement({
            sid: 'AllowRootUserToAdministerKey',
            effect: Effect.ALLOW,
            principals: [new AccountRootPrincipal()],
            actions: ['kms:*'],
            resources: resourceArns,
        }),
        denyEveryoneElseStatement,
    );

    policy.validateForResourcePolicy();

    return policy;
}
