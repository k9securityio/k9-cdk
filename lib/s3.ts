import * as s3 from "@aws-cdk/aws-s3";
import {BucketEncryption} from "@aws-cdk/aws-s3";
import {AccessCapability, AccessSpec, K9PolicyFactory} from "./k9policy";
import * as cdk from "@aws-cdk/core";
import {AddToResourcePolicyResult, AnyPrincipal, Effect, PolicyStatement} from "@aws-cdk/aws-iam";
import * as aws_iam_utils from "./aws-iam-utils";

export interface K9BucketPolicyProps extends s3.BucketPolicyProps {
    readonly k9DesiredAccess: Array<AccessSpec>
    readonly bucket: s3.Bucket
    readonly encryption?: BucketEncryption
}

let SUPPORTED_CAPABILITIES = new Array<AccessCapability>(
    AccessCapability.AdministerResource,
    AccessCapability.ReadConfig,
    AccessCapability.ReadData,
    AccessCapability.WriteData,
    AccessCapability.DeleteData,
);

export const SID_DENY_UNEXPECTED_ENCRYPTION_METHOD = 'DenyUnexpectedEncryptionMethod';

/**
 * Grants least-privilege access to a bucket by generating a BucketPolicy from the access capabilities
 * described by `props`; the policy will be set on the Bucket specified in `props`.
 * 
 * When a BucketPolicy already exists on the Bucket referenced in `props`:
 *   * the BucketPolicy's existing Statements will pass through unmodified
 *   * k9 will identify IAM principals there were allowed by the original policy and add those principals to
 *   the `DenyEveryoneElse` Statement's exclusion list so that, e.g. autoDeleteObjects works as expected
 *   * k9's Allow and Deny statements will be added to the policy
 *
 * @remarks
 *
 * k9 modifies the existing BucketPolicy in place instead of replacing or copying and modifying that
 * to preserve dependency references created by certain S3 CDK features such as `autoDeleteObjects`.
 *
 * @param scope The scope in which to define this construct.
 * @param id The scoped construct ID.
 * @param props describing the desired access capabilities for the bucket
 *
 * @return an array of AddToResourcePolicyResult
 */
export function grantAccessViaResourcePolicy(scope: cdk.Construct, id: string, props: K9BucketPolicyProps): AddToResourcePolicyResult[] {
    const policyFactory = new K9PolicyFactory();
    // If the bucket already has a policy, use it.  Maintaining the existing policy instance
    // is important because other CDK features like S3 autoDeleteObjects may have expressed dependencies
    // on that instance which must be maintained.
    if(!props.bucket.policy){
        props.bucket.policy = new s3.BucketPolicy(scope, `${id}Policy`, {bucket: props.bucket})
    }
    const policy = props.bucket.policy;

    const addToResourcePolicyResults = new Array<AddToResourcePolicyResult>();
    let resourceArns = [
        `${props.bucket.bucketArn}`,
        `${props.bucket.arnForObjects('*')}`
    ];

    // Capture the principals that were allowed prior to modifying policy
    // One could argue this can be done at the end because we're going to
    // narrow the DenyEveryoneElse to the unique set of allowed principals.
    // Record here for now to preserve ability to generate fine-grained DenyEveryoneElse-$capability statements.
    const origAllowedAWSPrincipals = aws_iam_utils.getAllowedPrincipalArns(policy.document);

    // Make Allow Statements
    const k9Statements = policyFactory.makeAllowStatements("S3",
        SUPPORTED_CAPABILITIES,
        props.k9DesiredAccess,
        resourceArns);

    // Make Deny Statements
    const denyEveryoneElseTest = policyFactory.wasLikeUsed(props.k9DesiredAccess) ?
        'ArnNotLike' :
        'ArnNotEquals';
    const denyEveryoneElseStatement = new PolicyStatement({
                sid: 'DenyEveryoneElse',
                effect: Effect.DENY,
                principals: policyFactory.makeDenyEveryoneElsePrincipals(),
                actions: ['s3:*'],
                resources: resourceArns
            });
    const allAllowedPrincipalArns = policyFactory.getAllowedPrincipalArns(props.k9DesiredAccess);
    for (let origAWSPrincipal of origAllowedAWSPrincipals){
        allAllowedPrincipalArns.add(origAWSPrincipal);
    }
    denyEveryoneElseStatement.addCondition(denyEveryoneElseTest,
        {'aws:PrincipalArn': [...allAllowedPrincipalArns]});

    // default encryption method to SSE-KMS,
    // allow override to SSE-S3 (AES256)
    let encryptionMethod = 'aws:kms'
    if(props.encryption){
        if(BucketEncryption.S3_MANAGED == props.encryption){
            encryptionMethod = 'AES256'
        }
    }
    k9Statements.push(
        new PolicyStatement({
            sid: 'DenyInsecureCommunications',
            effect: Effect.DENY,
            principals: [new AnyPrincipal()],
            actions: ['s3:*'],
            resources: resourceArns,
            conditions: {
                Bool: {'aws:SecureTransport': false},
            },
        }),
        new PolicyStatement({
            sid: 'DenyUnencryptedStorage',
            effect: Effect.DENY,
            principals: [new AnyPrincipal()],
            actions: ['s3:PutObject', 's3:ReplicateObject'],
            resources: resourceArns,
            conditions: {
                Null: {'s3:x-amz-server-side-encryption': true},
            },
        }),
        new PolicyStatement({
            sid: SID_DENY_UNEXPECTED_ENCRYPTION_METHOD,
            effect: Effect.DENY,
            principals: [new AnyPrincipal()],
            actions: ['s3:PutObject', 's3:ReplicateObject'],
            resources: resourceArns,
            conditions: {
                'StringNotEquals': {'s3:x-amz-server-side-encryption': encryptionMethod},
            },
        }),
        denyEveryoneElseStatement,
    );

    for(let statement of k9Statements){
        let addToResourcePolicyResult = props.bucket.addToResourcePolicy(statement);
        addToResourcePolicyResults.push(addToResourcePolicyResult);
    }

    policy.document.validateForResourcePolicy();

    return addToResourcePolicyResults;
}
