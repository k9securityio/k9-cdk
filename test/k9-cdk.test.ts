import {expect as expectCDK, haveResource} from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import {K9AccessCapabilities, K9BucketPolicyProps, K9PolicyFactory} from '../lib/k9-cdk-stack';

const administerResourceArns = new Set<string>([
        "arn:aws:iam::139710491120:user/ci",
    ]
);

const writeDataArns = new Set<string>([
        "arn:aws:iam::12345678910:role/app-backend",
    ]
);
const readDataArns = new Set<string>(writeDataArns)
    .add("arn:aws:iam::12345678910:role/customer-service");

const deleteDataArns = new Set<string>([
        "arn:aws:iam::139710491120:user/super-admin",
    ]
);

test('K9AccessCapabilities', () => {

    const k9AccessCapabilities: K9AccessCapabilities = {
        allowAdministerResourceArns: administerResourceArns,
        allowWriteDataArns: writeDataArns,
        allowReadDataArns: readDataArns,
        allowDeleteDataArns: deleteDataArns,
    };

    expect(k9AccessCapabilities).toEqual(k9AccessCapabilities);

    // Unit Test: k9 access capabilities
    expect(k9AccessCapabilities.allowAdministerResourceArns).toEqual(administerResourceArns);
    // expect(k9AccessCapabilities.allowAdministerResourceTest).toBe(ArnEqualsTest)
    expect(k9AccessCapabilities.allowWriteDataArns).toEqual(writeDataArns);
    expect(k9AccessCapabilities.allowReadDataArns).toEqual(readDataArns);
    expect(k9AccessCapabilities.allowDeleteDataArns).toEqual(deleteDataArns)

});

test('K9BucketPolicy', () => {
    const app = new cdk.App();

    const stack = new cdk.Stack(app, 'K9BucketPolicyTest');
    const bucket = new s3.Bucket(stack, 'TestBucket', {});

    const k9AccessCapabilities: K9AccessCapabilities = {
            allowAdministerResourceArns: administerResourceArns,
            allowWriteDataArns: writeDataArns,
            allowReadDataArns: readDataArns,
            // omit delete data ARNs and 'test' configs -- use ArnEquals
        };

    const k9BucketPolicyProps: K9BucketPolicyProps = {
        k9AccessCapabilities: k9AccessCapabilities,
        bucket: bucket
    };
    expect(k9BucketPolicyProps.k9AccessCapabilities).toEqual(k9AccessCapabilities);

    const k9PolicyFactory = new K9PolicyFactory();

    const bucketPolicy = k9PolicyFactory.makeBucketPolicy(stack, "S3Bucket", k9BucketPolicyProps);
    
    console.log("bucketPolicy: " + bucketPolicy.toString());
    console.log("bucketPolicy.document: " + JSON.stringify(bucketPolicy.document.toJSON(), null, 2));

    expectCDK(stack).to(haveResource("AWS::S3::Bucket"));
});
