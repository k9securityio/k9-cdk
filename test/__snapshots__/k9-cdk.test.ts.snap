// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`K9BucketPolicy 1`] = `
Object {
  "Resources": Object {
    "S3BucketPolicy189C1E8E": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestBucket560B80BC",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:BypassGovernanceRetention",
                "s3:DeleteBucketPolicy",
                "s3:DeleteBucketWebsite",
                "s3:PutAccelerateConfiguration",
                "s3:PutAnalyticsConfiguration",
                "s3:PutBucketAcl",
                "s3:PutBucketCORS",
                "s3:PutBucketLogging",
                "s3:PutBucketNotification",
                "s3:PutBucketObjectLockConfiguration",
                "s3:PutBucketPolicy",
                "s3:PutBucketPublicAccessBlock",
                "s3:PutBucketRequestPayment",
                "s3:PutBucketTagging",
                "s3:PutBucketVersioning",
                "s3:PutBucketWebsite",
                "s3:PutEncryptionConfiguration",
                "s3:PutInventoryConfiguration",
                "s3:PutLifecycleConfiguration",
                "s3:PutMetricsConfiguration",
                "s3:PutObjectAcl",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectVersionAcl",
                "s3:PutReplicationConfiguration",
              ],
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:PrincipalArn": Array [
                    "arn:aws:iam::139710491120:user/ci",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": "*",
              "Resource": "*",
              "Sid": "Restricted-administer-resource",
            },
            Object {
              "Action": Array [
                "s3:GetBucketAcl",
                "s3:GetBucketCORS",
                "s3:GetBucketLocation",
                "s3:GetBucketLogging",
                "s3:GetBucketNotification",
                "s3:GetBucketObjectLockConfiguration",
                "s3:GetBucketPolicy",
                "s3:GetBucketPolicyStatus",
                "s3:GetBucketPublicAccessBlock",
                "s3:GetBucketRequestPayment",
                "s3:GetBucketTagging",
                "s3:GetBucketVersioning",
                "s3:GetBucketWebsite",
                "s3:GetEncryptionConfiguration",
                "s3:GetInventoryConfiguration",
                "s3:GetLifecycleConfiguration",
                "s3:GetMetricsConfiguration",
                "s3:GetObject",
                "s3:GetObjectAcl",
                "s3:GetObjectLegalHold",
                "s3:GetObjectRetention",
                "s3:GetObjectTagging",
                "s3:GetObjectTorrent",
                "s3:GetObjectVersion",
                "s3:GetObjectVersionAcl",
                "s3:GetObjectVersionForReplication",
                "s3:GetObjectVersionTagging",
                "s3:GetObjectVersionTorrent",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads",
                "s3:ListBucketVersions",
                "s3:ListMultipartUploadParts",
              ],
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:PrincipalArn": Array [
                    "arn:aws:iam::12345678910:role/app-backend",
                    "arn:aws:iam::12345678910:role/customer-service",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": "*",
              "Resource": "*",
              "Sid": "Restricted-read-data",
            },
            Object {
              "Action": Array [
                "s3:AbortMultipartUpload",
                "s3:PutBucketTagging",
                "s3:PutObject",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:ReplicateDelete",
                "s3:ReplicateObject",
                "s3:ReplicateTags",
                "s3:RestoreObject",
              ],
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:PrincipalArn": Array [
                    "arn:aws:iam::12345678910:role/app-backend",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": "*",
              "Resource": "*",
              "Sid": "Restricted-write-data",
            },
            Object {
              "Action": Array [
                "s3:DeleteObject",
                "s3:DeleteObjectTagging",
                "s3:DeleteObjectVersion",
                "s3:DeleteObjectVersionTagging",
              ],
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:PrincipalArn": Array [],
                },
              },
              "Effect": "Allow",
              "Principal": "*",
              "Resource": "*",
              "Sid": "Restricted-delete-data",
            },
            Object {
              "Action": "*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": false,
                },
              },
              "Effect": "Deny",
              "Principal": "*",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestBucket560B80BC",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestBucket560B80BC": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
